### PR TITLE
Add action eventmaterialitemcrafts

### PR DESCRIPTION
**Pull request by tyrosine1153**
### Description
- Add action eventmaterialitemcrafts
- Add Sheet EventMaterialItemRecipeSheet

### Related Links
[아레나 그랜드 피날레 레시피 제작 개발](https://app.asana.com/0/958521740385861/1203307254532497/f)

### CODE REVIEW

File: Lib9c/Model/Mail/IMail.cs [hunk 1]
```diff
@@ -14,5 +14,7 @@ public interface IMail
         void Read(OrderBuyerMail orderBuyerMail);
         void Read(OrderSellerMail orderSellerMail);
         void Read(GrindingMail grindingMail);
+        void Read(MaterialCraftMail materialCraftMail);
+
```
Comment (sonohoshi) [inline Lib9c/Model/Mail/IMail.cs:18]:
```suggestion
```

File: Lib9c/TableData/Event/EventMaterialItemRecipeSheet.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public class EventMaterialItemRecipeSheet : Sheet<int, EventMaterialItemRecipeSheet.Row>
+    {
+        [Serializable]
+        public class Row : SheetRow<int>
+        {
+            public override int Key => Id;
+            public int Id { get; private set; }
+            public int ResultMaterialItemId { get; private set; }
+            public int ResultMaterialItemCount { get; private set; }
+            public int MaterialsCount { get; private set; }
```
Comment (sonohoshi) [inline Lib9c/TableData/Event/EventMaterialItemRecipeSheet.cs:17]:
```suggestion
            public int RequiredMaterialsCount { get; private set; }
```
How about it?

File: Lib9c/TableData/Event/EventMaterialItemRecipeSheet.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            public override void Set(IReadOnlyList<string> fields)
+            {
+                Id = ParseInt(fields[0]);
+                ResultMaterialItemId = ParseInt(fields[1]);
+                ResultMaterialItemCount = ParseInt(fields[2]);
+                MaterialsCount = ParseInt(fields[3]);
+                MaterialsId = new List<int>();
+                for (int i = 0; i < 12; i++)
+                {
+                    var id = fields[4 + i];
```
Comment (sonohoshi) [inline Lib9c/TableData/Event/EventMaterialItemRecipeSheet.cs:28]:
```suggestion
                const int maxCountOfMaterialTypes = 12;
                const int offset = 4;
                for (int i = 0; i < maxCountOfMaterialTypes ; i++)
                {
                    var id = fields[offset + i];
```

File: Lib9c/Model/Mail/MaterialCraftMail.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+using Nekoyume.Model.Item;
+using Nekoyume.Model.State;
+
+namespace Nekoyume.Model.Mail
+{
+    [Serializable]
+    public class MaterialCraftMail : Mail
+    {
+        public int ItemCount;
+        public Material Material;
```
Comment (sonohoshi) [inline Lib9c/Model/Mail/MaterialCraftMail.cs:14]: How about itemId instead of Material?

File: .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            {
+                level = 100,
+            };
+
+            _initialStates = _initialStates
+                .SetState(_agentAddress, agentState.Serialize())
+                .SetState(_avatarAddress, avatarState.SerializeV2())
+                .SetState(inventoryAddr, avatarState.inventory.Serialize())
+                .SetState(worldInformationAddr, avatarState.worldInformation.Serialize())
+                .SetState(questListAddr, avatarState.questList.Serialize()) // Todo : 이거 왜있지
```
Comment (sonohoshi) [inline .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs:67]:
```suggestion
                .SetState(questListAddr, avatarState.questList.Serialize())
```

File: .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+            Assert.Throws<InvalidMaterialCountException>(() =>
+            {
+                Execute(
+                    _initialStates,
+                    eventScheduleId,
+                    eventMaterialItemRecipeId,
+                    materialsToUse,
+                    scheduleRow.StartBlockIndex);
+            });
+        }
```
Comment (sonohoshi) [inline .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs:213]: The name is `Execute_InvalidAddressException`, but why do you check `InvalidMaterialCountException`?

File: .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+            {
+                PreviousStates = previousStates,
+                Signer = _agentAddress,
+                Random = new TestRandom(),
+                Rehearsal = false,
+                BlockIndex = blockIndex,
+            });
+
+            var nextAvatarState = nextStates.GetAvatarStateV2(_avatarAddress);
+            Assert.Equal(
```
Comment (sonohoshi) [inline .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs:270]: How about checking consumed material?

File: .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs (hunk 6 above)
Comment (tyrosine1153) [reply .Lib9c.Tests/Action/EventMaterialItemCraftsTest.cs:213]: Oh...........that's a typo

File: Lib9c/Action/EventMaterialItemCrafts.cs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+                        avatarState.questList.Serialize());
+            }
+            else
+            {
+                states = states
+                    .SetState(AvatarAddress, avatarState.SerializeV2())
+                    .SetState(
+                        AvatarAddress.Derive(LegacyInventoryKey),
+                        avatarState.inventory.Serialize());
+            }
```
Comment (ipdae) [inline Lib9c/Action/EventMaterialItemCrafts.cs:259]: inventory, avatarState always serialize without migrationRequired flags.

File: Lib9c/Action/EventMaterialItemCrafts.cs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            sw.Stop();
+            Log.Verbose(
+                "[{ActionTypeString}][{AddressesHex}] Set states: {Elapsed}",
+                ActionTypeText,
+                addressesHex,
+                sw.Elapsed);
+            // ~Set states
+
+            Log.Verbose(
```
Comment (ipdae) [inline Lib9c/Action/EventMaterialItemCrafts.cs:268]: change log level to debug start and executed total time.

File: Lib9c/Extensions/EventRecipeExtensions.cs [hunk 10]
```diff
@@ -44,5 +45,77 @@ public static class EventRecipeExtensions
...
+                    actionTypeText,
+                    addressesHex,
+                    recipeRow.RequiredMaterialsCount,
+                    materialsCount);
+            }
+        }
+
+        public static List<EventMaterialItemRecipeSheet.Row> GetRecipeRows(
+            this EventMaterialItemRecipeSheet sheet,
+            int eventScheduleId)
```
Comment (ipdae) [inline Lib9c/Extensions/EventRecipeExtensions.cs:108]: where is this method used?

**Review by sonohoshi — APPROVED** [review]

File: Lib9c/Extensions/EventRecipeExtensions.cs (hunk 10 above)
Comment (tyrosine1153) [reply Lib9c/Extensions/EventRecipeExtensions.cs:108]: It used in 9c client.

Comment (ipdae) [reply Lib9c/Extensions/EventRecipeExtensions.cs:108]: If it's only used on the client, it would be better to move this method to the client side.

**Review by boscohyun — APPROVED** [review]

**Pull request merged**